### PR TITLE
Fix: remove catastrophic backtracking vulnerability

### PR DESCRIPTION
**Pull request by davisjam**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Change template substitution regex to exclude fields with whitespace.
This addresses possible O(n^2) catastrophic backtracking behavior.
The function behavior is unchanged.

Very unlikely to be exploited. For #10002.

**Is there anything you'd like reviewers to focus on?**

No.

### CODE REVIEW

File: lib/util/interpolate.js [hunk 1]
```diff
@@ -13,7 +13,12 @@ module.exports = (text, data) => {
     if (!data) {
         return text;
     }
-    return text.replace(/\{\{\s*([^{}]+?)\s*\}\}/g, (fullMatch, term) => {
+
+    // Substitution content for any {{ }} markers.
+    return text.replace(/\{\{([^{}]+?)\}\}/g, (fullMatch, term) => {
+
+        // Strip leading and trailing whitespace.
+        term = term.replace(/^\s+|\s+$/g, "");
```
Comment (ljharb) [inline lib/util/interpolate.js]: `term.trim()`?

Comment (davisjam) [reply lib/util/interpolate.js]: Haha! I knew there had to be a function for that.

**Comment by davisjam** [comment]
> So it's just the \s* that's problematic here?

A bit more complex than that. The original pattern contains a sub-sequence like this:

```js
\s*[^{}]+\s*
```
Because the middle group's character class includes whitespace, it will match a superset of this pattern, which is easier to reason about:

```js
\s*\s+\s*
```

If the input contains a long sequence of whitespace, the regex engine has to decide when to move from one \s to the next.
If the input doesn't match the pattern, the regex engine tries every subdivision of the whitespace looking for a match. It lays breadcrumbs each time it makes a decision to resolve ambiguity, and revisiting each breadcrumb is called backtracking.

**Comment by davisjam** [comment]
It would be helpful if a developer can comment on whether this is a potential low-risk security problem.
If so I will follow up with Snyk.io to get a low-severity vulnerability ID assigned.

**Pull request merged**